README
This is TclRAL a relational algebra library extention for Tcl.

INSTALLATION
============

Installation follows the typical TEA scheme. On most UNIX or Linux systems
it should be sufficient to unwrap the source and:

% ./configure
% make
% make install

Configure typically needs to know where Tcl is installed on your system
via the "--with-tcl=...." argument. You may also wish to remove the
assertions by defining "CPPFLAGS=-DNDEBUG" as an environment variable.

This also works more or less on Windows system using MSys or Cygwin.
I build the Windows binary distributes using MSys.

BINARY DISTRIBUTION
===================

As of Revision 0.9.0, binary distributions are in the form or zip package
suitable for installation using the "teacup" command.  It also works to unzip
the package file in an appropriate place along your Tcl library path.  This
could be in the "lib" directory of your Tcl installation (a common practice
under Windows) or in any directory that is mentioned in the TCLLIBPATH
environment variable. Binary distributions come with documentation and test
cases located in directories below the package files.

CONTACT
=======

TclRAL is a hosted at SourceForge and bug reports and requests can be posted
there:
    http://sourceforge.net/projects/tclral


CHANGES
========= ral 0.11.2 =========
2014/08/03
    * Procedural constraints are now evaluated in the namespace implied
      by the fully qualified constraint name.
    * The commands are now a proper ensemble with a mapping dictionary.
    * Corrected a number of problems in the SQLite serialization and
      added mergeToSQLite.

========= ral 0.10.2 =========
2012/02/26 mangoa01

	* configure:
	* configure.in:
	* doc/ral/relation.man:
	* doc/ral/relformat.man:
	* doc/ral/relvar.man:
	* doc/ral/serialize.man:
	* doc/ral/tuple.man:
	* ralutil/doc/ralutil.man:
	* ralutil/pkgIndex.tcl:
	* ralutil/ralutil.tcl:
	* ralutil/teapot.txt:
	* src/ral.c:
	* src/ral.tcl:
	* src/ral_attribute.c:
	* src/ral_relation.c:
	* src/ral_relationcmd.c:
	* src/ral_relvar.c:
	* src/ral_relvarcmd.c:
	* src/ral_relvarobj.c:
	* src/ral_tuplecmd.c:
	* src/ral_tupleheading.c:
	* test/ral.test:
	  Fixed the bignum problem by properly initializing the Tom Math stubs.
	  Cleaned up a number of compiler warning for unused variables when
	  compiled with NDEBUG to remove assertions.

========= ral 0.10.1 =========
2011/10/08 mangoa01

	* doc/ral/relation.man:
	* doc/ral/relformat.man:
	* doc/ral/relvar.man:
	* doc/ral/serialize.man:
	* doc/ral/tuple.man:
	* ralutil/doc/ralutil.man:
	  Update documentation version numbers in preparation for new release.

	* ralutil/pkgIndex.tcl:
	* ralutil/ralutil.tcl:
	* ralutil/teapot.txt:
	  Update to ralutil package revision numbers.
          Improved system generated ID's procedures. Now the path name to a
          relvar will be automatically resolved and if the attribute that is
          the system generated ID is missing in the trace callback, it is added
          automatically.

2011/10/03 mangoa01

	* src/ral_relation.c:
          Put in a workaround for Tcl 8.4. The hash table code in 8.4 clears
          the ClientData field whenever any new entry is created -- even if a
          custom hash type supplied an allocation function. Later versions of
          the code in Tcl 8.5 do not clear ClientData if the allocation
          function was supplied (i.e. it depends upon that allocation function
          to set a value into ClientData). This difference causes the NULL
          pointer to dereferenced when TclRAL is compiled against Tcl 8.4. Here
          we put in a workaround.

2011/09/26 mangoa01

	* src/ral.tcl:
	* src/ral_relationcmd.c:
	* src/ral_relvar.h:
	* src/ral_relvarcmd.c:
	* src/ral_relvarobj.c:
	* src/ral_utils.h:
	* test/ral.test:
	  Corrected several issues when trying to compile for Tcl 8.4 and for
	  compilers that are strict about C89 syntax.

2011/08/16 mangoa01

	* configure:
	* configure.in:
	* src/ral_relvarobj.c:
	  Corrected a problem with a format statement that contained too few
	  arguments.
	  Incremented version number in preparation for a new release.

	* src/ral.tcl:
	  Increment revision number in Tcl script portion of the package.

2011/08/09 mangoa01

	* src/ral_relation.c:
	  Remove C99 style comment for the benefit of older compilers.

========= ral 0.10.0 =========
2011/06/08 mangoa01

	* ralutil/teapot.txt:
	* src/ral.tcl:
	  Corrected a problem with foreign key constraints when serializing
	  relvars to a SQLite database.

2011/06/06 mangoa01

	* doc/ral/relation.man:
	* doc/ral/relformat.man:
	* doc/ral/relvar-examples.tcl:
	* doc/ral/relvar.man:
	* doc/ral/serialize.man:
	* doc/ral/tuple.man:
	* ralutil/doc/ralutil.man:
	* src/ral_relationcmd.c:
	* src/ral_relvarobj.c:
	  Clean up of documenation, adding the remaining relvar commands that
	  are going into the 0.10.0 release.

2011/06/05 mangoa01

	* Makefile.in:
	* configure:
	* configure.in:
	* src/ral.c:
	* src/ral.tcl:
	* src/ral_attribute.c:
	* src/ral_attribute.h:
	* src/ral_joinmap.c:
	* src/ral_joinmap.h:
	* src/ral_relation.c:
	* src/ral_relation.h:
	* src/ral_relationcmd.c:
	* src/ral_relationcmd.h:
	* src/ral_relationobj.c:
	* src/ral_relationobj.h:
	* src/ral_relvar.c:
	* src/ral_relvar.h:
	* src/ral_relvarcmd.c:
	* src/ral_relvarcmd.h:
	* src/ral_relvarobj.c:
	* src/ral_relvarobj.h:
	* src/ral_tuple.c:
	* src/ral_tuple.h:
	* src/ral_tuplecmd.c:
	* src/ral_tuplecmd.h:
	* src/ral_tupleheading.c:
	* src/ral_tupleheading.h:
	* src/ral_tupleobj.c:
	* src/ral_tupleobj.h:
	* src/ral_utils.c:
	* src/ral_utils.h:
	* src/ral_vector.c:
	* src/ral_vector.h:
	* tclconfig/install-sh:
	* tclconfig/tcl.m4:
	* test/ral.test:
	  Completed procedural constrain command.
	  Updated to latest TEA files (3.9).
          Corrected long latent bug where duplicated attribute names in a
          relation join "-using" argument caused a panic and aborted the
          program.

2011/05/30 mangoa01

	* doc/ral/relation.man:
	* test/ral.test:
	  Updates for "uinsert" command.

2011/04/03 mangoa01

	* src/ral.tcl:
	* src/ral_relvar.c:
	* src/ral_relvar.h:
	* src/ral_relvarcmd.c:
	* src/ral_relvarobj.c:
	* src/ral_relvarobj.h:
	* src/ral_utils.c:
	* src/ral_utils.h:
	  Checkpoint. Saving the current state of the procedural constraint
	  code. It's not done yet, but I wanted it in the repository.

2011/01/16 mangoa01

	* configure:
	* configure.in:
	* doc/ral/relation.man:
	* doc/ral/relformat.man:
	* doc/ral/relvar.man:
	* doc/ral/serialize.man:
	* doc/ral/tuple.man:
	* ralutil/doc/ralutil.man:
	* ralutil/pkgIndex.tcl:
	* ralutil/ralutil.tcl:
	* src/ral_relation.c:
	* src/ral_relation.h:
	* src/ral_relationcmd.c:
	* src/ral_relvar.c:
	* src/ral_relvarcmd.c:
	* src/ral_utils.c:
	* src/ral_utils.h:
	* test/ral.test:
	* test/serialize.test:
	  Added commands to do disjoin union of both relation values and
	  of values stored in relvars.

2010/12/05 mangoa01

	* src/ral.tcl:
	  Completed SQLite load / store procedures.

	* src/ral.tcl:
	* src/ral_relationcmd.c:
	  First cut at proc to store and load sets of relvars to/from SQLite.

2010/11/14 mangoa01

	* src/ral_relationcmd.c:
          Added "relation uinsert" command to perform an insert without
          throwing an error on duplicates.


src/ral_utils.c
/*
This software is copyrighted 2005 - 2011 by G. Andrew Mangogna.  The following
terms apply to all files associated with the software unless explicitly
disclaimed in individual files.

The authors hereby grant permission to use, copy, modify, distribute,
and license this software and its documentation for any purpose, provided
that existing copyright notices are retained in all copies and that this
notice is included verbatim in any distributions. No written agreement,
license, or royalty fee is required for any of the authorized uses.
Modifications to this software may be copyrighted by their authors and
need not follow the licensing terms described here, provided that the
new terms are clearly indicated on the first page of each file where
they apply.

IN NO EVENT SHALL THE AUTHORS OR DISTRIBUTORS BE LIABLE TO ANY PARTY FOR
DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OF THIS SOFTWARE, ITS DOCUMENTATION, OR ANY DERIVATIVES
THEREOF, EVEN IF THE AUTHORS HAVE BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.

THE AUTHORS AND DISTRIBUTORS SPECIFICALLY DISCLAIM ANY WARRANTIES,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE, AND NON-INFRINGEMENT.  THIS SOFTWARE
IS PROVIDED ON AN "AS IS" BASIS, AND THE AUTHORS AND DISTRIBUTORS HAVE
NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS,
OR MODIFICATIONS.

GOVERNMENT USE: If you are acquiring this software on behalf of the
U.S. government, the Government shall have only "Restricted Rights"
in the software and related documentation as defined in the Federal
Acquisition Regulations (FARs) in Clause 52.227.19 (c) (2).  If you
are acquiring the software on behalf of the Department of Defense,
the software shall be classified as "Commercial Computer Software"
and the Government shall have only "Restricted Rights" as defined in
Clause 252.227-7013 (c) (1) of DFARs.  Notwithstanding the foregoing,
the authors grant the U.S. Government and others acting in its behalf
permission to use and distribute the software in accordance with the
terms specified in this license.
*/
/*
 *++
MODULE:

ABSTRACT:

$RCSfile: ral_utils.c,v $
$Revision: 1.26 $
$Date: 2011/06/05 18:01:10 $
 *--
 */

/*
PRAGMAS
*/

/*
INCLUDE FILES
*/
#include "ral_utils.h"
#include <string.h>
#include <assert.h>

/*
MACRO DEFINITIONS
*/

/*
TYPE DEFINITIONS
*/

/*
EXTERNAL FUNCTION REFERENCES
*/

/*
FORWARD FUNCTION REFERENCES
*/

/*
EXTERNAL DATA REFERENCES
*/

/*
EXTERNAL DATA DEFINITIONS
*/

/*
STATIC DATA ALLOCATION
*/

static char const * const cmdStrings[] = {
    "unknown command",
    "tuple",
    "relation",
    "relvar",
    "setfromany",
    "updatefromobj",
} ;

/*
 * The order here must match that of the definition of "enum Ral_CmdOption"!
 */
static char const * const optStrings[] = {
    "NONE",
    "array",
    "assign",
    "association",
    "body",
    "cardinality",
    "compose",
    "constraint",
    "correlation",
    "create",
    "degree",
    "delete",
    "deleteone",
    "destroy",
    "dict",
    "divide",
    "dunion",
    "eliminate",
    "emptyof",
    "equal",
    "eval",
    "extend",
    "extract",
    "foreach",
    "fromdict",
    "fromlist",
    "get",
    "group",
    "heading",
    "identifiers",
    "insert",
    "intersect",
    "is",
    "isempty",
    "isnotempty",
    "issametype",
    "join",
    "list",
    "minus",
    "names",
    "partition",
    "procedural",
    "project",
    "rank",
    "relation",
    "rename",
    "restrict",
    "restrictone",
    "restrictwith",
    "semijoin",
    "semiminus",
    "set",
    "summarize",
    "summarizeby",
    "table",
    "tag",
    "tclose",
    "times",
    "trace",
    "tuple",
    "uinsert",
    "ungroup",
    "union",
    "unset",
    "unwrap",
    "update",
    "updateone",
    "updateper",
    "wrap",
} ;

/*
 * The order here must match that of the definition of "enum Ral_ErrorCode"!
 */
static char const * const resultStrings[] = {
    "no error",
    "unknown attribute name",
    "duplicate attribute name",
    "bad heading format",
    "bad value format",
    "bad value type for value",
    "unknown data type",
    "bad type keyword",
    "wrong number of attributes specified",
    "bad list of pairs",
    "duplicate command option",

    "relations of non-zero degree must have at least one identifier",
    "identifiers must have at least one attribute",
    "identifiers must not be subsets of other identifiers",
    "tuple has duplicate values for an identifier",
    "duplicate attribute name in identifier attribute set",
    "duplicate tuple",
    "headings not equal",
    "relation must have degree of one",
    "relation must have degree of two",
    "relation must have cardinality of one",
    "bad list of triples",
    "attributes do not constitute an identifier",
    "attribute must be of a Relation type",
    "attribute must be of a Tuple type",
    "relation is not a projection of the summarized relation",
    "divisor heading must be disjoint from the dividend heading",
    "mediator heading must be a union of the dividend and divisor headings",
    "too many attributes specified",
    "attributes must have the same type",
    "only a single identifier may be specified",
    "identifier must have only a single attribute",
    "\"-within\" option attributes are not the subset of any identifier",
    "attribute is not a valid type for rank operation",

    "duplicate relvar name",
    "unknown relvar name",
    "mismatch between referential attributes",
    "duplicate constraint name",
    "unknown constraint name",
    "relvar has constraints in place",
    "referred to identifiers can not have non-singular multiplicities",
    "operation is not allowed during \"eval\" command",
    "a super set relvar may not be named as one of its own sub sets",
    "correlation spec is not available for a \"-complete\" correlation",
    "recursively invoking a relvar command outside of a transaction",
    "recursive attempt to modify a relvar already being changed",

    "serious internal error",
} ;

static char const * const errorStrings[] = {
    "OK",
    "UNKNOWN_ATTR",
    "DUPLICATE_ATTR",
    "HEADING_ERR",
    "FORMAT_ERR",
    "BAD_VALUE",
    "BAD_TYPE",
    "BAD_KEYWORD",
    "WRONG_NUM_ATTRS",
    "BAD_PAIRS_LIST",
    "DUPLICATE_OPTION",

    "NO_IDENTIFIER",
    "IDENTIFIER_FORMAT",
    "IDENTIFIER_SUBSET",
    "IDENTITY_CONSTRAINT",
    "DUP_ATTR_IN_ID",
    "DUPLICATE_TUPLE",
    "HEADING_NOT_EQUAL",
    "DEGREE_ONE",
    "DEGREE_TWO",
    "CARDINALITY_ONE",
    "BAD_TRIPLE_LIST",
    "NOT_AN_IDENTIFIER",
    "NOT_A_RELATION",
    "NOT_A_TUPLE",
    "NOT_A_PROJECTION",
    "NOT_DISJOINT",
    "NOT_UNION",
    "TOO_MANY_ATTRS",
    "TYPE_MISMATCH",
    "SINGLE_IDENTIFIER",
    "SINGLE_ATTRIBUTE",
    "WITHIN_NOT_SUBSET",
    "BAD_RANK_TYPE",

    "DUP_NAME",
    "UNKNOWN_NAME",
    "REFATTR_MISMATCH",
    "DUP_CONSTRAINT",
    "UNKNOWN_CONSTRAINT",
    "CONSTRAINTS_PRESENT",
    "BAD_MULT",
    "BAD_TRANS_OP",
    "SUPER_NAME",
    "INCOMPLETE_SPEC",
    "ONGOING_CMD",
    "ONGOING_MODIFICATION",

    "INTERNAL_ERROR",
} ;
/*
FUNCTION DEFINITIONS
*/

void
Ral_ErrorInfoSetCmd(
    Ral_ErrorInfo *info,
    Ral_Command cmd,
    Ral_CmdOption opt)
{
    if (info) {
	info->cmd = cmd ;
	info->opt = opt ;
        info->errorCode = RAL_ERR_OK ;
    }
}

void
Ral_ErrorInfoSetError(
    Ral_ErrorInfo *info,
    Ral_ErrorCode errorCode,
    const char *param)
{
    if (info) {
	info->errorCode = errorCode ;
	Tcl_DStringInit(&info->param) ;
	Tcl_DStringAppend(&info->param, param, -1) ;
    }
}

void
Ral_ErrorInfoSetErrorObj(
    Ral_ErrorInfo *info,
    Ral_ErrorCode errorCode,
    Tcl_Obj *objPtr)
{
    if (info) {
	info->errorCode = errorCode ;
	Tcl_DStringInit(&info->param) ;
	Tcl_DStringAppend(&info->param, Tcl_GetString(objPtr), -1) ;
    }
}

void
Ral_InterpSetError(
    Tcl_Interp *interp,
    Ral_ErrorInfo *info)
{
    if (interp && info) {
	const char *param ;

        assert(info->errorCode < RAL_ERR_LAST_ERRORCODE) ;

	param = Tcl_DStringValue(&info->param) ;
	Tcl_ResetResult(interp) ;
	Tcl_AppendStringsToObj(Tcl_GetObjResult(interp),
	    resultStrings[info->errorCode], ", \"", param, "\"", NULL) ;
	Tcl_SetErrorCode(interp, "RAL", cmdStrings[info->cmd],
	    optStrings[info->opt], errorStrings[info->errorCode], param, NULL) ;
    }
    if (info) {
	Tcl_DStringFree(&info->param) ;
    }
}

void
Ral_InterpErrorInfo(
    Tcl_Interp *interp,
    Ral_Command cmd,
    Ral_CmdOption opt,
    Ral_ErrorCode errorCode,
    const char *param)
{
    Ral_ErrorInfo errInfo ;

    Ral_ErrorInfoSetCmd(&errInfo, cmd, opt) ;
    Ral_ErrorInfoSetError(&errInfo, errorCode, param) ;
    Ral_InterpSetError(interp, &errInfo) ;
}

void
Ral_InterpErrorInfoObj(
    Tcl_Interp *interp,
    Ral_Command cmd,
    Ral_CmdOption opt,
    Ral_ErrorCode errorCode,
    Tcl_Obj *objPtr)
{
    Ral_ErrorInfo errInfo ;

    Ral_ErrorInfoSetCmd(&errInfo, cmd, opt) ;
    Ral_ErrorInfoSetErrorObj(&errInfo, errorCode, objPtr) ;
    Ral_InterpSetError(interp, &errInfo) ;
}

char const *
Ral_ErrorInfoGetCommand(
    Ral_ErrorInfo *info)
{
    return cmdStrings[info->cmd] ;
}

char const *
Ral_ErrorInfoGetOption(
    Ral_ErrorInfo *info)
{
    return optStrings[info->opt] ;
}
